url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/fri/patterns/interpreter/parsergenerator/examples/XmlLexer.java

```diff
@@ -0,0 +1,125 @@
+package fri.patterns.interpreter.parsergenerator.examples;
+
+import java.util.*;
+import java.io.*;
+import fri.util.TimeStopper;
+import fri.util.io.UnicodeReader;
+import fri.patterns.interpreter.parsergenerator.builder.SerializedLexer;
+import fri.patterns.interpreter.parsergenerator.lexer.LexerImpl;
+import fri.patterns.interpreter.parsergenerator.lexer.LexerSemantic;
+import fri.patterns.interpreter.parsergenerator.lexer.ResultTree;
+import fri.patterns.interpreter.parsergenerator.syntax.Rule;
+
+/**
+	Example XML lexer. Not event-driven like SAX, but good for DOM-building.
+	
+	@author Fritz Ritzberger, 2003
+*/
+
+public class XmlLexer
+{
+	public static void main(String [] args)
+		throws Exception
+	{
+		if (args.length <= 0)	{
+			System.err.println("SYNTAX: java "+XmlLexer.class.getName()+" file.xml [file.xml ...]");
+			System.err.println("	Example XML Parser");
+			System.exit(1);
+		}
+
+		// Standalone lexer as top-down parser.
+		TimeStopper timer = new TimeStopper();
+		// Building lexer from scratch takes 840 millis. Parsing takes 60 millis for a 70 line XML file.
+
+		// read the syntax from EBNF file
+		Reader syntaxInput = new InputStreamReader(XmlLexer.class.getResourceAsStream("Xml.syntax"));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *syntaxInput*. Currently, there are execution paths that do not contain closure statements, for example, when SerializedLexer.get throws an exception. Ensure *syntaxInput* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).